### PR TITLE
Added euler path for Graphs, Digraphs and Multigraphs

### PR DESCRIPTION
**Pull request by edwardo2013**
We added the file find_euler_path.py which has an implementation of the euler path algorithm. It is general: It works for Graphs, Digraphs and Multigraphs. We added some tests after the function. It uses the eulerian_circuit() function, which is implemented in euler.py . We make the comments following the aforementioned file.

Please tell us is this is the way to document the function. If not, provide information/files to add the required documentation.

About the code: the function receives a networkx graph G. If G is eulerian, then just returns the eulerian circuit. If not, then verify degree condition of nodes to see if it has an euler path. If it does not have it, raise an error. Then, at line 145 the code to find the euler path is given (this works because we know that we have the conditions to have an euler path). It returns a generator which has the euler path. 

Please review this code, we would be happy to see it merged in the networkx project. Please let us know if some changes must be made (emails are in the file) or if it is posible to make the code more compact.

Thanks,
Edwardo S. Rivera

### CODE REVIEW

File: networkx/algorithms/find_euler_path.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+import networkx as nx
+
+""" 
+Find an eulerian path
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: "Eulerian" is derived from someone's name; it should be capitalized everywhere.

File: networkx/algorithms/find_euler_path.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            # we have neighbors, so add the vertex to the stack (2), take any of its neighbors (1)
+            # remove the edge between selected neighbor and that vertex,
+            # and set that neighbor as the current vertex (4).
+            else:
+                random_edge = next(edges(current_vertex)) #(1)
+                vertex_stack.append(get_vertex(random_edge)) #(2)
+                g.remove_edge(*random_edge) #(3)
+
+# tests
+if __name__ == "__main__":
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Tests should be in a test module, like `tests/test_euler.py` not in the module containing the function.

File: networkx/algorithms/find_euler_path.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+#    Humberto Ortiz <humberto@hpcf.upr.edu>
+#    BSD license. 
+#    All rights reserved.
+
+
+__author__ = """\n""".join(['Edwardo Rivera <edwardo.rivera@urp.edu>','Humberto Ortiz <humberto@hpcf.upr.edu>'])
+
+__all__ = ['find_eulerian_path'] 
+
+def find_eulerian_path(G):
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: This should be in `euler.py`, not in its own module.

File: networkx/algorithms/find_euler_path.py (hunk 1 above)
Comment (edwardo2013) [inline networkx/algorithms/find_euler_path.py]: Will be fixed, thanks.

File: networkx/algorithms/find_euler_path.py (hunk 2 above)
Comment (edwardo2013) [inline networkx/algorithms/find_euler_path.py]: Thanks. Is there some specific way to do the testing for networkX?

Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: See, for example, https://github.com/networkx/networkx/blob/master/networkx/algorithms/tests/test_euler.py. Create new test methods similar to, for example, `test_eulerian_circuit_digraph`.

Comment (edwardo2013) [inline networkx/algorithms/find_euler_path.py]: Ok, thanks for the information.

File: networkx/algorithms/find_euler_path.py (hunk 3 above)
Comment (edwardo2013) [inline networkx/algorithms/find_euler_path.py]: So, how I should write the authors?  I don't want to touch code from someone else :)

Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: I'm not sure what the purpose of the `__author__` variable is; it seems to be an artifact of a time before GitHub. If its purpose is simply to provide contact information for authors of substantial portions of the code, then it makes sense to add your contact information there.

**Comment by hagberg** [comment]
Re: __author__ . Yes those were the days before git and github and use of this variable is optional.

File: networkx/algorithms/find_euler_path.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    G.add_edge("N","E")
+    G.add_edge("E","S")
+    G.add_edge("S","W")
+    G.add_edge("W","E")
+    x = find_eulerian_path(G)
+    nodepath = [u for u, v in x]
+    print nodepath
+
+    G = nx.DiGraph([(1,2),(2,3)])
+    print list(find_eulerian_path(G))
```
Comment (SanketDG) [inline networkx/algorithms/find_euler_path.py]: This isn't unit testing. Google "python unit testing" to see how they are done. `assert` is used extensively to check for the desired output, nothing is printed in testing.

Comment (edwardo2013) [inline networkx/algorithms/find_euler_path.py]: I would use nose to do the testing (as in euler.py tests).

Comment (SanketDG) [inline networkx/algorithms/find_euler_path.py]: Yes, this library uses `nose` but every testing framework follows the rules from python unit testing. Look at other tests for a reference.

**Comment by edwardo2013** [comment]
As soon as I am done with college, I will finish this. Thanks for all the input!

File: networkx/algorithms/find_euler_path.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')])
+    >>> list(nx.find_eulerian_path(G))
+    [('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')]
+
+    >>> G = nx.Digraph([(1, 2), (2, 3)])
+    >>> list(nx.find_eulerian_path(G))
+    [(1,2),(2,3)]
+    """
+    from operator import itemgetter
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Imports must come at the head of the file.

File: networkx/algorithms/find_euler_path.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    if G.is_directed() and not nx.is_weakly_connected(G):
+        raise nx.NetworkXError("G is not connected.")    
+
+    # is undirected
+    if not G.is_directed() and not nx.is_connected(G):
+        raise nx.NetworkXError("G is not connected.")
+
+    # Now verify if has an eulerian circuit: even condition of all nodes is satified.
+    if nx.is_eulerian(G):
+        x = nx.eulerian_circuit(G) # generator of edges
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Why not just `return nx.eulerian_circuit(G)`?

File: networkx/algorithms/find_euler_path.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        for i in x:
+            yield i
+        
+    # Not all vertex have even degree, check if exactly two vertex have odd degrees.
+    # If yes, then there is an Euler path. If not, raise an error (no euler path can be found)
+    else:
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: The `else` is redundant here, remove it and dedent its body.

File: networkx/algorithms/find_euler_path.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        g = G.__class__(G)  # copy graph structure (not attributes)
+        
+        # list to check the odd degree condition, and a flag
+        check_odd = []
+        directed = False
+    
+        if g.is_directed():
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Better: `directed = g.is_directed(); if directed: ...`

File: networkx/algorithms/find_euler_path.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            edges = g.in_edges_iter
+            get_vertex = itemgetter(0)
+            directed = True
+        else:
+            degree = g.degree
+            edges = g.edges_iter
+            get_vertex = itemgetter(1)
+            
+        # Verify if an euler path can be found. Complexity O(n) ?
+        for vertex in g.nodes():
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Simpler: `for v in G:`

File: networkx/algorithms/find_euler_path.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+            deg = degree(vertex)
+            # directed case
+            if directed:
+                outdeg = out_degree(vertex)
+                if deg != outdeg:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                # is odd and we append it. 
+                    else:
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Redundant `else`, remove it.

Also fix the comment on the previous line.

File: networkx/algorithms/find_euler_path.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+                        check_odd.append(vertex)
+            # undirected case        
+            else:
+                if deg % 2 != 0:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                    # is odd and we append it. 
+                    else:
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Redundant `else`, remove it.

File: networkx/algorithms/find_euler_path.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+                        check_odd.append(vertex)
+
+        if directed:
+            def verify_odd_cond(g,check_odd):
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Why is this its own function? Just do the checks directly without defining a new function.

File: networkx/algorithms/find_euler_path.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+                first = check_odd[0]
+                second = check_odd[1]
+                if  g.out_degree(first) == g.in_degree(first) + 1 and g.in_degree(second) == g.out_degree(second) + 1:
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Line's too long.

File: networkx/algorithms/find_euler_path.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+                    return second
+                elif g.out_degree(second) == g.in_degree(second) + 1 and g.in_degree(first)  == g.out_degree(first) + 1:
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Redundant `elif`, replace it with `if`.

File: networkx/algorithms/find_euler_path.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+                    return None
+            start = verify_odd_cond(g,check_odd)
+        else:
+            start = check_odd[0]
+
+        # if the odd condition is not meet, raise an error.
+        if not start:
+            raise nx.NetworkXError("G doesn't have an Euler Path")
+        # Begin algorithm: 
+        vertex_stack = [start]
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Probably should use `collections.deque` here if possible.

File: networkx/algorithms/find_euler_path.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+        last_vertex = None
+
+        while vertex_stack:
+
+            current_vertex = vertex_stack[-1] #(4)
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: Remove these inline comments that are just numbers.

File: networkx/algorithms/find_euler_path.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+                # Special case, we cannot add a None vertex to the path.
+                if last_vertex is not None:
+                    yield (last_vertex, current_vertex)
+                last_vertex = current_vertex
+                vertex_stack.pop()
+            # we have neighbors, so add the vertex to the stack (2), take any of its neighbors (1)
+            # remove the edge between selected neighbor and that vertex,
+            # and set that neighbor as the current vertex (4).
+            else:
+                random_edge = next(edges(current_vertex)) #(1)
```
Comment (jfinkels) [inline networkx/algorithms/find_euler_path.py]: This is not a "random" edge, it is an "arbitrary" edge. Better to just name the variable `edge`.

File: networkx/algorithms/find_euler_path.py (hunk 6 above)
Comment (humberto-ortiz) [inline networkx/algorithms/find_euler_path.py]: The function is a generator, and you can't return a value. We have to yield the edges.

By the way, I'm cleaning up these issues and will submit a new pull request from my account.

Edwardo graduated.

**Comment by dschult** [comment]
See #1878

**Pull request closed without merging**